File: 8. String to Integer (atoi).cpp
class Solution {
public:
    int myAtoi(string str) {
        int n = str.length();
        int res = 0;
        int new_num;        
        int flag = 0;
        if (n == 0)
            return 0;
        for (int i = 0; i < n; i++){
            if (str[i] == ' ' && flag == 0)
                continue;
            else if (str[i] >= '0' && str[i] <= '9'){                
                if (flag == 0)
                    flag = 1;
                new_num = str[i] - '0';
                if (flag && (INT_MIN + new_num) * 1.0 / 10 > res)
                    return INT_MIN;
                else if ((INT_MAX - new_num) * 1.0 / 10 < res)
                    return INT_MAX;     
                if (flag == 1)
                    res = res * 10 + new_num;                
                else
                    res = res * 10 - new_num;
            }        
            else if (str[i] == '+' && flag == 0)
                flag = 1;
            else if (str[i] == '-' && flag == 0)
                flag = -1;
            else
                return res;
        }
        return res;
    }
};

File: 5-Problem-A-Day/8. String to Integer (atoi).java
class Solution {
    public int myAtoi(String str) {
        int flag = 1;
        int ans = 0;
        boolean findNum = false;
        for (char c : str.toCharArray()) {
            if (c == ' ' && !findNum) 
                continue;            
            else if (c == '-' && !findNum){
                flag = -1;
                findNum = true;
            }
            else if (c == '+' && !findNum) {
                flag = 1;
                findNum = true;
            }
            else if (c >= '0' && c <= '9') {
                findNum = true;
                if (ans > 0 && ans > (Integer.MAX_VALUE - c + '0') / 10)
                    return Integer.MAX_VALUE;
                else if (ans < 0 && ans < (Integer.MIN_VALUE + c - '0') / 10)
                    return Integer.MIN_VALUE;                
                ans = ans * 10 + flag * (c - '0');
            }
            else
                return ans;
        }
        return ans;
    }
}
